oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -571,20 +576,23 @@ impl Runtime {
     fn validate_can_read_from_label(
         &self,
         node_id: NodeId,
-        label: &Label,
+        source_label: &Label,
     ) -> Result<(), OakStatus> {
-        let downgraded_node_label = self.get_node_downgraded_label(node_id);
+        // We downgrade the source Label, since its data is being read into the current Node.
+        let downgraded_source_label = self.get_node_downgraded_label(node_id, source_label);
+        let target_label = self.get_node_label(node_id);
+        trace!("{:?}: original source label: {:?}?", node_id, source_label);
         trace!(
-            "{:?}: can {:?} read from {:?}?",
+            "{:?}: downgraded source label: {:?}?",
             node_id,
-            downgraded_node_label,
-            label
+            downgraded_source_label
         );
-        if label.flows_to(&downgraded_node_label) {
-            trace!("{:?}: can read from {:?}", node_id, label);
+        trace!("{:?}: target label: {:?}?", node_id, target_label);
+        if downgraded_source_label.flows_to(&target_label) {
+            trace!("{:?}: can read from {:?}", node_id, source_label);
```

Comment:
Idle musing: I wonder if it will end up being helpful for debugging to distinguish in the logs between cases where downgrade is and isn't required.  
```Rust
let target_label = self.get_node_label(node_id);
if source_label.flows_to(&target_label) {
  trace!("{:?}: can read from {:?}, => {:?}", node_id, source_label, target_label);
  Ok(())
} else {
  // See if a downgraded version of the source Label can be read into the current Node.
  let downgraded_source_label = self.get_node_downgraded_label(node_id, source_label);
  if downgraded_source_label.flows_to(&target_label) {
    trace!("{:?}: can read from downgraded {:?}, => {:?}", node_id, downgraded_source_label, target_label);
    Ok(())
  } else {
    debug!("{:?}: cannot read from {:?}, =/=> {:?}", node_id, source_label, target_label);
    Err(OakStatus::PermissionDenied)
  }
}

Reply:
I second this musing, but not a big deal.

Reply:
I also mused about it, I would like to clean up other parts of the logs too, I created https://github.com/project-oak/oak/issues/1052 to track (also I wanted to try out the "reference in new issue" functionality from GitHub)